NMS-10268  render dateformat for Vaadin (including user time zone)

patrick-schweizer opened the pull request:
* JIRA: https://issues.opennms.org/browse/NMS-10268

Our [continuous integration system](http://bamboo.internal.opennms.com:8085) will test and verify your changes.

patrick-schweizer commented:
replaces: https://github.com/OpenNMS/opennms/pull/2071

patrick-schweizer commented:
@mvrueden , @j-white  I rebased the PR on foudation-2018 - it should be ready for review now

mvrueden commented on features/vaadin-dashboard/src/main/java/org/opennms/features/vaadin/dashboard/model/TimeformatHelper.java:
```diff
@@ -0,0 +1,45 @@
+
+import com.vaadin.server.VaadinSession;
+
+public final class TimeformatHelper {
```
Can we rename this? 
`TimeformatHelper` sounds like something which would help me formatting times, but instead it helps me getting the user's time zone.

patrick-schweizer replied:
done, renamed to UserTimeZoneExtractor

patrick-schweizer commented on features/timeformat/api/pom.xml:
```diff
@@ -0,0 +1,36 @@
+      </plugin>
+    </plugins>
+  </build>
+  <dependencies>
```
done.

patrick-schweizer commented on features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java:
```diff
@@ -76,20 +78,26 @@
         m_nodeDao = nodeDao;
         m_resourceDao = resourceDao;
         m_transactionOperations = transactionOperations;
+        this.timeformatService = timeformatService;
```
done, replaced m_ by this.

patrick-schweizer commented:
@mvrueden I rebased the PR to the current foundation-2018, the user timezone settings are in there:
https://github.com/OpenNMS/opennms/blob/jira/NMS-10268-with-NMS-10228/opennms-webapp/src/main/webapp/admin/userGroupView/users/modifyUser.jsp#L456
https://github.com/OpenNMS/opennms/blob/jira/NMS-10268-with-NMS-10228/opennms-webapp/src/main/java/org/opennms/web/account/UserSettingsFilter.java

mvrueden commented:
Originally I only commented on one line about not mixing `m_` and `this`, and only that was addressed here. It would be nice to not mix both techniques in general.

Rule of thumb: If there is a new member variable, use the already existing scheme. Otherwise we end up having a lot of to the issue unrelated changes which makes possible backports much harder.
The same is true for formatting changes and also re-arranging imports.

Besides that code wise it looks good.

When testing this, I noticed, that not all vaadin applications use the new format.
For example, the SNMP Mib Compiler shows the default toString() representation: `Thu Aug 02 15:37:34 CEST 2018`
The same is true for the `Last Event Time` in the Topology Map (and Node Maps).

mvrueden commented:
One side note, Bamboo is having troubles, so the `Create Packages` stage failed. As soon as bamboo is working again, we may have to manually trigger a new build

patrick-schweizer commented on features/topology-map/plugins/org.opennms.features.topology.plugins.browsers/src/main/java/org/opennms/features/topology/plugins/browsers/TimeColumnGenerator.java:
```diff
@@ -0,0 +1,79 @@
+        return formattedValue;
+    }
+
+    // TODO: replace method with UserTimeZoneExtractor
```
@mvrueden what would be the best place for UserTimeZoneExtractor so it can be accessed from TimeColumnGenerator and the Dashlets? Thanks!

patrick-schweizer commented:
@mvrueden I addressed your comments:
* not mixing m_ and this: fixed
* SNMP Mib Compiler shows the default toString() representation: fixed
* Topology Map: fixed
* Node Maps: fixed

open:
* I am not sure if there is another place in the Vaadin code that does Date formatting?
* I am not sure where to put UserTimeZoneExtractor (see comment above)

patrick-schweizer commented:
@mvrueden I addressed your comments:
* not mixing m_ and this: fixed
* SNMP Mib Compiler shows the default toString() representation: fixed
* Topology Map: fixed
* Node Maps: fixed

open:
* I am not sure if there is another place in the Vaadin code that does Date formatting?
* I am not sure where to put UserTimeZoneExtractor in order to be available to all Vaadin code (see comment above)

mvrueden commented:
Originally the idea was to put components shared through all vaadin applications in `features/vaadin-components`, so I'd start looking if it could live there.

I have to go through the code myself to see if we covered all places accordingly.

patrick-schweizer commented:
@mvrueden , thanks, I moved UserTimeZoneExtractor to features/vaadin-components/core.
I went through the code and tried to identify all places with time formatting (including tables defined in blueprint.xml) but it could very well be that I missed a place.

mvrueden commented:
I asked @christianpape to also take a look (see PR https://github.com/OpenNMS/opennms/pull/2101 for more details) and he mentioned that "his" vaadin applications are migrated. Given that and what I saw I think we are good to go from there.

Besides that if there are any places we missed, it will be easy now to migrate it to the "new way" of formatting times/dates.

In addition, @RangerRick had some concerns about the `@Deprecated` methods in the newly created `TimeService`. In today's sprint meeting we agreed on addressing this before merging.

mvrueden commented on features/vaadin-dashlets/dashlet-ksc/src/main/java/org/opennms/features/vaadin/dashboard/dashlets/KscDashlet.java:
```diff
@@ -72,30 +74,36 @@
     public DashletComponent getWallboardComponent() {
-        if (m_wallboardComponent == null) {
-            m_wallboardComponent = new AbstractDashletComponent() {
+        if (wallboardComponent == null) {
```
This was probably changed by mistake?

patrick-schweizer commented:
@mvrueden as sgreed with @RangerRick I removed the deprecated annotation since it does a nullcheck on the date object.
I also fixed the missed m_ vs. this. variables

=> if bamboo is green I hope we are good to merge

RangerRick commented:
Those last changes look reasonable to me.  @j-white you're technically still "awaiting review" here but I'm cool with merging this whenever.
